src/router/busybox/docs/BusyBox.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.3
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "BUSYBOX 1"
.TH BUSYBOX 1 "2006-05-28" "version 1.1.3" "BusyBox"
.SH "NAME"
BusyBox \- The Swiss Army Knife of Embedded Linux
.SH "SYNTAX"
.IX Header "SYNTAX"
.Vb 1
\& BusyBox <function> [arguments...]  # or
.Ve
.PP
.Vb 1
\& <function> [arguments...]          # if symlinked
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
BusyBox combines tiny versions of many common \s-1UNIX\s0 utilities into a single
small executable. It provides minimalist replacements for most of the utilities
you usually find in \s-1GNU\s0 coreutils, util\-linux, etc. The utilities in BusyBox
generally have fewer options than their full-featured \s-1GNU\s0 cousins; however, the
options that are included provide the expected functionality and behave very
much like their \s-1GNU\s0 counterparts.
.PP
BusyBox has been written with size-optimization and limited resources in mind.
It is also extremely modular so you can easily include or exclude commands (or
features) at compile time. This makes it easy to customize your embedded
systems. To create a working system, just add /dev, /etc, and a Linux kernel.
BusyBox provides a fairly complete \s-1POSIX\s0 environment for any small or embedded
system.
.PP
BusyBox is extremely configurable.  This allows you to include only the
components you need, thereby reducing binary size. Run 'make config' or 'make
menuconfig' to select the functionality that you wish to enable.  Then run
\&'make' to compile BusyBox using your configuration.
.PP
After the compile has finished, you should use 'make install' to install
BusyBox. This will install the 'bin/busybox' binary, in the target directory
specified by \s-1PREFIX\s0. \s-1PREFIX\s0 can be set when configuring BusyBox, or you can
specify an alternative location at install time (i.e., with a command line
like 'make PREFIX=/tmp/foo install'). If you enabled any applet installation
scheme (either as symlinks or hardlinks), these will also be installed in
the location pointed to by \s-1PREFIX\s0.
.SH "USAGE"
.IX Header "USAGE"
BusyBox is a multi-call binary.  A multi-call binary is an executable program
that performs the same job as more than one utility program.  That means there
is just a single BusyBox binary, but that single binary acts like a large
number of utilities.  This allows BusyBox to be smaller since all the built-in
utility programs (we call them applets) can share code for many common operations.
.PP
You can also invoke BusyBox by issuing a command as an argument on the
command line.  For example, entering
.PP
.Vb 1
\&        /bin/busybox ls
.Ve
.PP
will also cause BusyBox to behave as 'ls'.
.PP
Of course, adding '/bin/busybox' into every command would be painful.  So most
people will invoke BusyBox using links to the BusyBox binary.
.PP
For example, entering
.PP
.Vb 2
\&        ln -s /bin/busybox ls
\&        ./ls
.Ve
.PP
will cause BusyBox to behave as 'ls' (if the 'ls' command has been compiled
into BusyBox).  Generally speaking, you should never need to make all these
links yourself, as the BusyBox build system will do this for you when you run
the 'make install' command.
.PP
If you invoke BusyBox with no arguments, it will provide you with a list of the
applets that have been compiled into your BusyBox binary.
.SH "COMMON OPTIONS"
.IX Header "COMMON OPTIONS"
Most BusyBox commands support the \fB\-\-help\fR argument to provide a terse runtime
description of their behavior.  If the \s-1CONFIG_FEATURE_VERBOSE_USAGE\s0 option has
been enabled, more detailed usage information will also be available.
.SH "COMMANDS"
.IX Header "COMMANDS"
Currently defined functions include:
.PP
.Vb 28
\&        [, [[, addgroup, adduser, adjtimex, ar, arping, ash, awk, 
\&        basename, bbconfig, bunzip2, busybox, bzcat, cal, cat, chattr, 
\&        chgrp, chmod, chown, chroot, chvt, clear, cmp, comm, cp, cpio, 
\&        crond, crontab, cut, date, dc, dd, deallocvt, delgroup, deluser, 
\&        devfsd, df, dirname, dmesg, dnsd, dos2unix, dpkg, dpkg_deb, du, 
\&        dumpkmap, dumpleases, e2fsck, echo, eject, env, ether_wake, expr, 
\&        fakeidentd, false, fbset, fdflush, fdformat, fdisk, find, fold, 
\&        free, freeramdisk, fsck, fsck_minix, ftpget, ftpput, fuser, 
\&        getopt, getty, grep, gunzip, gzip, halt, hdparm, head, hexdump, 
\&        hostid, hostname, httpd, hwclock, id, ifconfig, ifdown, ifup, 
\&        inetd, init, insmod, install, ip, ipaddr, ipcalc, ipcrm, ipcs, 
\&        iplink, iproute, iptunnel, kill, killall, klogd, lash, last, 
\&        length, less, ln, loadfont, loadkmap, logger, login, logname, 
\&        logread, losetup, ls, lsattr, lsmod, lzmacat, makedevs, md5sum, 
\&        mdev, mesg, mkdir, mke2fs, mkfifo, mkfs_minix, mknod, mkswap, 
\&        mktemp, modprobe, more, mount, mountpoint, mt, mv, nameif, nc, 
\&        netstat, nice, nohup, nslookup, od, openvt, passwd, patch, pidof, 
\&        ping, ping6, pivot_root, poweroff, printenv, printf, ps, pwd, 
\&        rdate, readlink, readprofile, realpath, reboot, renice, reset, 
\&        rm, rmdir, rmmod, route, rpm, rpm2cpio, run_parts, runlevel, rx, 
\&        sed, seq, setarch, setconsole, setkeycodes, setsid, sha1sum, 
\&        sleep, sort, start_stop_daemon, stat, strings, stty, su, sulogin, 
\&        sum, swapoff, swapon, switch_root, sync, sysctl, syslogd, tail, 
\&        tar, tee, telnet, telnetd, test, tftp, time, top, touch, tr, 
\&        traceroute, true, tty, tune2fs, udhcpc, udhcpd, umount, uname, 
\&        uncompress, uniq, unix2dos, unlzma, unzip, uptime, usleep, 
\&        uudecode, uuencode, vconfig, vi, vlock, watch, watchdog, wc, 
\&        wget, which, who, whoami, xargs, yes, zcat, zcip
.Ve
.SH "COMMAND DESCRIPTIONS"
.IX Header "COMMAND DESCRIPTIONS"
.IP "\fBaddgroup  \fR" 4
.IX Item "addgroup  "
addgroup   [\fB\-g\fR \s-1GID\s0] group_name [user_name]
.Sp
Adds a group to the system
.Sp
Options:
.Sp
.Vb 1
\&        -g GID          specify gid
.Ve
.IP "\fBadduser   \fR" 4
.IX Item "adduser   "
adduser    [\s-1OPTIONS\s0] user_name
.Sp
Adds a user to the system
.Sp
Options:
.Sp
.Vb 7
\&        -h DIR          Assign home directory DIR
\&        -g GECOS        Assign gecos field GECOS
\&        -s SHELL        Assign login shell SHELL
\&        -G              Add the user to existing group GROUP
\&        -S              create a system user (ignored)
\&        -D              Do not assign a password (logins still possible via ssh)
\&        -H              Do not create the home directory
.Ve
.IP "\fBadjtimex  \fR" 4
.IX Item "adjtimex  "
adjtimex   [\fB\-q\fR] [\fB\-o\fR offset] [\fB\-f\fR frequency] [\fB\-p\fR timeconstant] [\fB\-t\fR tick]
.Sp
Reads and optionally sets system timebase parameters.
See \fIadjtimex\fR\|(2).
.Sp
Options:
.Sp
.Vb 6
\&        -q              quiet mode - do not print
\&        -o offset       time offset, microseconds
\&        -f frequency    frequency adjust, integer kernel units (65536 is 1ppm)
\&                        (positive values make the system clock run fast)
\&        -t tick         microseconds per tick, usually 10000
\&        -p timeconstant
.Ve
.IP "\fBar        \fR" 4
.IX Item "ar        "
ar         [\fB\-o\fR] [\fB\-v\fR] [\fB\-p\fR] [\fB\-t\fR] [\fB\-x\fR] \s-1ARCHIVE\s0 \s-1FILES\s0
.Sp
Extract or list \s-1FILES\s0 from an ar archive.
.Sp
Options:
.Sp
.Vb 5
\&        -o              preserve original dates
\&        -p              extract to stdout
\&        -t              list
\&        -x              extract
\&        -v              verbosely list files processed
.Ve
.IP "\fBarping    \fR" 4
.IX Item "arping    "
arping     [\fB\-fqbDUA\fR] [\fB\-c\fR count] [\fB\-w\fR timeout] [\fB\-I\fR device] [\fB\-s\fR sender] target
.Sp
Ping hosts by \s-1ARP\s0 requests/replies.
.Sp
Options:
.Sp
.Vb 11
\&        -f              Quit on first ARP reply
\&        -q              Be quiet
\&        -b              Keep broadcasting, don't go unicast
\&        -D              Duplicated address detection mode
\&        -U              Unsolicited ARP mode, update your neighbours
\&        -A              ARP answer mode, update your neighbours
\&        -c count        Stop after sending count ARP request packets
\&        -w timeout      Time to wait for ARP reply, in seconds
\&        -I device       Outgoing interface name, default is eth0
\&        -s sender       Set specific sender IP address
\&        target          Target IP address of ARP request
.Ve
.IP "\fBash       \fR" 4
.IX Item "ash       "
ash        [\s-1FILE\s0]...
or: ash \fB\-c\fR command [args]...
.Sp
The ash shell (command interpreter)
.IP "\fBawk       \fR" 4
.IX Item "awk       "
awk        [\s-1OPTION\s0]... [program\-text] [\s-1FILE\s0 ...]
.Sp
Options:
.Sp
.Vb 3
\&        -v var=val      assign value 'val' to variable 'var'
\&        -F sep          use 'sep' as field separator
\&        -f progname     read program source from file 'progname'
.Ve
.IP "\fBbasename  \fR" 4
.IX Item "basename  "
basename   \s-1FILE\s0 [\s-1SUFFIX\s0]
.Sp
Strips directory path and suffixes from \s-1FILE\s0.
If specified, also removes any trailing \s-1SUFFIX\s0.
.Sp
Example:
.Sp
.Vb 6
\&        $ basename /usr/local/bin/foo
\&        foo
\&        $ basename /usr/local/bin/
\&        bin
\&        $ basename /foo/bar.txt .txt
\&        bar
.Ve
.IP "\fBbbconfig  \fR" 4
.IX Item "bbconfig  "
bbconfig   	
.Sp
Print the config file which built busybox
.IP "\fBbunzip2   \fR" 4
.IX Item "bunzip2   "
bunzip2    [\s-1OPTION\s0]... [\s-1FILE\s0]
.Sp
Uncompress \s-1FILE\s0 (or standard input if \s-1FILE\s0 is '\-' or omitted).
.Sp
Options:
.Sp
.Vb 2
\&        -c      Write output to standard output
\&        -f      Force
.Ve
.IP "\fBbusybox   \fR" 4
.IX Item "busybox   "
busybox    
.Sp
Hello world!
.IP "\fBbzcat     \fR" 4
.IX Item "bzcat     "
bzcat      \s-1FILE\s0
.Sp
Uncompress to stdout.
.IP "\fBcal       \fR" 4
.IX Item "cal       "
cal        [\fB\-jy\fR] [[month] year]
.Sp
Display a calendar.
.Sp
Options:
.Sp
.Vb 2
\&        -j      Use julian dates
\&        -y      Display the entire year
.Ve
.IP "\fBcat       \fR" 4
.IX Item "cat       "
cat        [\fB\-u\fR] [\s-1FILE\s0]...
.Sp
Concatenates \s-1FILE\s0(s) and prints them to stdout.
.Sp
Options:
.Sp
.Vb 1
\&        -u      ignored since unbuffered i/o is always used
.Ve
.Sp
Example:
.Sp
.Vb 2
\&        $ cat /proc/uptime
\&        110716.72 17.67
.Ve
.IP "\fBchattr    \fR" 4
.IX Item "chattr    "
chattr     [\fB\-R\fR] [\-+=AacDdijsStTu] [\fB\-v\fR version] files...
.Sp
change file attributes on an ext2 fs
.Sp
Modifiers:
.Sp
.Vb 4
\&        -       remove attributes
\&        +       add attributes
\&        =       set attributes
\&Attributes:
.Ve
.Sp
.Vb 12
\&        A       don't track atime
\&        a       append mode only
\&        c       enable compress
\&        D       write dir contents synchronously
\&        d       do not backup with dump
\&        i       cannot be modified (immutable)
\&        j       write all data to journal first
\&        s       zero disk storage when deleted
\&        S       write file contents synchronously
\&        t       disable tail-merging of partial blocks with other files
\&        u       allow file to be undeleted
\&Options:
.Ve
.Sp
.Vb 2
\&        -R      recursively list subdirectories
\&        -v      set the file's version/generation number
.Ve
.IP "\fBchgrp     \fR" 4
.IX Item "chgrp     "
chgrp      [\s-1OPTION\s0]... \s-1GROUP\s0 \s-1FILE\s0...
.Sp
Change the group membership of each \s-1FILE\s0 to \s-1GROUP\s0.
.Sp
Options:
.Sp
.Vb 1
\&        -R      Changes files and directories recursively
.Ve
.Sp
Example:
.Sp
.Vb 5
\&        $ ls -l /tmp/foo
\&        -r--r--r--    1 andersen andersen        0 Apr 12 18:25 /tmp/foo
\&        $ chgrp root /tmp/foo
\&        $ ls -l /tmp/foo
\&        -r--r--r--    1 andersen root            0 Apr 12 18:25 /tmp/foo
.Ve
.IP "\fBchmod     \fR" 4
.IX Item "chmod     "
chmod      [\fB\-R\fR] MODE[,MODE]... \s-1FILE\s0...
.Sp
Each \s-1MODE\s0 is one or more of the letters ugoa, one of the
symbols +\-= and one or more of the letters rwxst.
.Sp
Options:
.Sp
.Vb 1
\&        -R      Changes files and directories recursively
.Ve
.Sp
Example:
.Sp
.Vb 8
\&        $ ls -l /tmp/foo
\&        -rw-rw-r--    1 root     root            0 Apr 12 18:25 /tmp/foo
\&        $ chmod u+x /tmp/foo
\&        $ ls -l /tmp/foo
\&        -rwxrw-r--    1 root     root            0 Apr 12 18:25 /tmp/foo*
\&        $ chmod 444 /tmp/foo
\&        $ ls -l /tmp/foo
\&        -r--r--r--    1 root     root            0 Apr 12 18:25 /tmp/foo
.Ve
.IP "\fBchown     \fR" 4
.IX Item "chown     "
chown      [ \fB\-Rh\fR ]...  OWNER[<.|:>[\s-1GROUP\s0]] \s-1FILE\s0...
.Sp
Change the owner and/or group of each \s-1FILE\s0 to \s-1OWNER\s0 and/or \s-1GROUP\s0.
.Sp
Options:
.Sp
.Vb 2
\&        -R      Changes files and directories recursively
\&        -h      Do not dereference symbolic links
.Ve
.Sp
Example:
.Sp
.Vb 8
\&        $ ls -l /tmp/foo
\&        -r--r--r--    1 andersen andersen        0 Apr 12 18:25 /tmp/foo
\&        $ chown root /tmp/foo
\&        $ ls -l /tmp/foo
\&        -r--r--r--    1 root     andersen        0 Apr 12 18:25 /tmp/foo
\&        $ chown root.root /tmp/foo
\&        ls -l /tmp/foo
\&        -r--r--r--    1 root     root            0 Apr 12 18:25 /tmp/foo
.Ve
.IP "\fBchroot    \fR" 4
.IX Item "chroot    "
chroot     \s-1NEWROOT\s0 [\s-1COMMAND\s0...]
.Sp
Run \s-1COMMAND\s0 with root directory set to \s-1NEWROOT\s0.
.Sp
Example:
.Sp
.Vb 6
\&        $ ls -l /bin/ls
\&        lrwxrwxrwx    1 root     root          12 Apr 13 00:46 /bin/ls -> /BusyBox
\&        # mount /dev/hdc1 /mnt -t minix
\&        # chroot /mnt
\&        # ls -l /bin/ls
\&        -rwxr-xr-x    1 root     root        40816 Feb  5 07:45 /bin/ls*
.Ve
.IP "\fBchvt      \fR" 4
.IX Item "chvt      "
chvt       N
.Sp
Changes the foreground virtual terminal to /dev/ttyN
.IP "\fBclear     \fR" 4
.IX Item "clear     "
clear      	
.Sp
Clear screen.
.IP "\fBcmp       \fR" 4
.IX Item "cmp       "
cmp        [\fB\-l\fR] [\fB\-s\fR] \s-1FILE1\s0 [\s-1FILE2\s0]
.Sp
Compares \s-1FILE1\s0 vs stdin if \s-1FILE2\s0 is not specified.
.Sp
Options:
.Sp
.Vb 3
\&        -l      Write the byte numbers (decimal) and values (octal)
\&                  for all differing bytes
\&        -s      quiet mode - do not print
.Ve
.IP "\fBcomm      \fR" 4
.IX Item "comm      "
comm       [\fB\-123\fR] \s-1FILE1\s0 \s-1FILE2\s0
.Sp
Compares \s-1FILE1\s0 to \s-1FILE2\s0, or to stdin if = is specified.
.Sp
Options:
.Sp
.Vb 3
\&        -1      Suppress lines unique to FILE1
\&        -2      Suppress lines unique to FILE2
\&        -3      Suppress lines common to both files
.Ve
.IP "\fBcp        \fR" 4
.IX Item "cp        "
cp         [\s-1OPTION\s0]... \s-1SOURCE\s0 \s-1DEST\s0
.Sp
Copies \s-1SOURCE\s0 to \s-1DEST\s0, or multiple \s-1SOURCE\s0(s) to \s-1DIRECTORY\s0.
.Sp
.Vb 7
\&        -a      Same as -dpR
\&        -d,-P   Preserves links
\&        -H,-L   Dereference all symlinks (implied by default)
\&        -p      Preserves file attributes if possible
\&        -f      force (implied; ignored) - always set
\&        -i      interactive, prompt before overwrite
\&        -R,-r   Copies directories recursively
.Ve
.IP "\fBcpio      \fR" 4
.IX Item "cpio      "
cpio       \-[dimtuv][F cpiofile]
.Sp
Extract or list files from a cpio archive
Main operation mode:
.Sp
.Vb 7
\&        d               make leading directories
\&        i               extract
\&        m               preserve mtime
\&        t               list
\&        v               verbose
\&        u               unconditional overwrite
\&        F               input from file
.Ve
.IP "\fBcrond     \fR" 4
.IX Item "crond     "
crond      \fB\-d\fR[#] \fB\-c\fR <crondir> \fB\-f\fR \fB\-b\fR
.Sp
.Vb 8
\&        -d [#] -l [#] -S -L logfile -f -b -c dir
\&        -d num  debug level
\&        -l num  log level (8 - default)
\&        -S      log to syslogd (default)
\&        -L file log to file
\&        -f      run in fordeground
\&        -b      run in background (default)
\&        -c dir  working dir
.Ve
.IP "\fBcrontab   \fR" 4
.IX Item "crontab   "
crontab    [\fB\-c\fR dir] {file|\-}|[\fB\-u\fR|\fB\-l\fR|\fB\-e\fR|\fB\-d\fR user]
.Sp
.Vb 7
\&        file <opts>  replace crontab from file
\&        -    <opts>  replace crontab from stdin
\&        -u user      specify user
\&        -l [user]    list crontab for user
\&        -e [user]    edit crontab for user
\&        -d [user]    delete crontab for user
\&        -c dir       specify crontab directory
.Ve
.IP "\fBcut       \fR" 4
.IX Item "cut       "
cut        [\s-1OPTION\s0]... [\s-1FILE\s0]...
.Sp
Prints selected fields from each input \s-1FILE\s0 to standard output.
.Sp
Options:
.Sp
.Vb 6
\&        -b LIST         Output only bytes from LIST
\&        -c LIST         Output only characters from LIST
\&        -d CHAR         Use CHAR instead of tab as the field delimiter
\&        -s              Output only the lines containing delimiter
\&        -f N            Print only these fields
\&        -n              Ignored
.Ve
.Sp
Example:
.Sp
.Vb 4
\&        $ echo "Hello world" | cut -f 1 -d ' '
\&        Hello
\&        $ echo "Hello world" | cut -f 2 -d ' '
\&        world
.Ve
.IP "\fBdate      \fR" 4
.IX Item "date      "
date       [\s-1OPTION\s0]... [MMDDhhmm[[\s-1CC\s0]YY][.ss]] [+FORMAT]
.Sp
Displays the current time in the given \s-1FORMAT\s0, or sets the system date.
.Sp
Options:
.Sp
.Vb 10
\&        -R              Outputs RFC-822 compliant date string
\&        -d STRING       Displays time described by STRING, not `now'
\&        -I[TIMESPEC]    Outputs an ISO-8601 compliant date/time string
\&                        TIMESPEC=`date' (or missing) for date only,
\&                        `hours', `minutes', or `seconds' for date and,
\&                        time to the indicated precision
\&        -D hint         Use 'hint' as date format, via strptime()
\&        -s              Sets time described by STRING
\&        -r FILE         Displays the last modification time of FILE
\&        -u              Prints or sets Coordinated Universal Time
.Ve
.Sp
Example:
.Sp
.Vb 2
\&        $ date
\&        Wed Apr 12 18:52:41 MDT 2000
.Ve
.IP "\fBdc        \fR" 4
.IX Item "dc        "
dc         expression ...
.Sp
This is a Tiny \s-1RPN\s0 calculator that understands the
following operations: +, add, \-, sub, *, mul, /, div, %, mod, **, exp, and, or, not, eor.
For example: 'dc 2 2 add' \-> 4, and 'dc 8 8 \e* 2 2 + /' \-> 16.
.Sp
Options:
p \- Prints the value on the top of the stack, without altering the stack
f \- Prints the entire contents of the stack without altering anything
o \- Pops the value off the top of the stack and uses it to set the output radix
.Sp
.Vb 1
\&    Only 10 and 16 are supported
.Ve
.Sp
Example:
.Sp
.Vb 10
\&        $ dc 2 2 + p
\&        4
\&        $ dc 8 8 \e* 2 2 + / p
\&        16
\&        $ dc 0 1 and p
\&        0
\&        $ dc 0 1 or p
\&        1
\&        $ echo 72 9 div 8 mul p | dc
\&        64
.Ve
.IP "\fBdd        \fR" 4
.IX Item "dd        "
dd         [if=FILE] [of=FILE] [bs=N] [count=N] [skip=N]
	  [seek=N] [conv=notrunc|noerror|sync]
.Sp
Copy a file, converting and formatting according to options
.Sp
.Vb 9
\&        if=FILE         read from FILE instead of stdin
\&        of=FILE         write to FILE instead of stdout
\&        bs=N            read and write N bytes at a time
\&        count=N         copy only N input blocks
\&        skip=N          skip N input blocks
\&        seek=N          skip N output blocks
\&        conv=notrunc    don't truncate output file
\&        conv=noerror    continue after read errors
\&        conv=sync       pad blocks with zeros
.Ve
.Sp
Numbers may be suffixed by c (x1), w (x2), b (x512), kD (x1000), k (x1024),
\&\s-1MD\s0 (x1000000), M (x1048576), \s-1GD\s0 (x1000000000) or G (x1073741824)
.Sp
Example:
.Sp
.Vb 3
\&        $ dd if=/dev/zero of=/dev/ram1 bs=1M count=4
\&        4+0 records in
\&        4+0 records out
.Ve
.IP "\fBdeallocvt \fR" 4
.IX Item "deallocvt "
deallocvt  [N]
.Sp
Deallocate unused virtual terminal /dev/ttyN
.IP "\fBdelgroup  \fR" 4
.IX Item "delgroup  "
delgroup   \s-1GROUP\s0
.Sp

src/router/busybox/applets/busybox.c
/* vi: set sw=4 ts=4: */
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <stdlib.h>
#include "busybox.h"
#if ENABLE_LOCALE_SUPPORT
#include <locale.h>
#else
#define setlocale(x,y)
#endif
#include <rc.h>

const char *bb_applet_name ATTRIBUTE_EXTERNALLY_VISIBLE;

#ifdef CONFIG_FEATURE_INSTALLER
/*
 * directory table
 *		this should be consistent w/ the enum, busybox.h::Location,
 *		or else...
 */
static const char usr_bin [] ="/usr/bin";
static const char usr_sbin[] ="/usr/sbin";

static const char* const install_dir[] = {
	&usr_bin [8], /* "", equivalent to "/" for concat_path_file() */
	&usr_bin [4], /* "/bin" */
	&usr_sbin[4], /* "/sbin" */
	usr_bin,
	usr_sbin
};

/* abstract link() */
typedef int (*__link_f)(const char *, const char *);

/* create (sym)links for each applet */
static void install_links(const char *busybox, int use_symbolic_links)
{
	__link_f Link = link;

	char *fpc;
	int i;
	int rc;

	if (use_symbolic_links)
		Link = symlink;

	for (i = 0; applets[i].name != NULL; i++) {
		fpc = concat_path_file(
			install_dir[applets[i].location], applets[i].name);
		rc = Link(busybox, fpc);
		if (rc!=0 && errno!=EEXIST) {
			bb_perror_msg("%s", fpc);
		}
		free(fpc);
	}
}

#else
#define install_links(x,y)
#endif /* CONFIG_FEATURE_INSTALLER */

int main(int argc, char **argv)
{
  char *base = strrchr (argv[0], '/');
      
  base = base ? base + 1 : argv[0];

#ifdef HAVE_DHCPFWD
  if (strstr (base, "dhcpfwd"))
    return dhcpforward_main(argc,argv);
#else
  if (strstr (base, "blahblah"))
    return puts(base);
#endif
/*#ifdef HAVE_PPP
  else if (strstr (base, "pppd"))
    return pppd_main(argc,argv);
#endif*/
#ifdef HAVE_RFLOW
  else if (strstr (base, "rflow"))
    return rflow_main(argc,argv);
#endif
#ifdef HAVE_SSHD
  else if (strstr (base, "dropbearkey"))
    return dropbearkey_main(argc,argv);
  else if (strstr (base, "dropbearkonvert"))
    return dropbearconvert_main(argc,argv);
  else if (strstr (base, "dropbear"))
    return dropbear_main(argc,argv);
  else if (strstr (base, "dbclient"))
    return cli_main(argc,argv);
  else if (strstr (base, "ssh"))
    return cli_main(argc,argv);    
  else if (strstr (base, "scp"))
    return scp_main(argc,argv);
#endif
#ifdef HAVE_BIRD
  else if (strstr (base, "bird"))
    return bird_main(argc, argv);
#endif    
  else if (strstr (base, "dnsmasq"))
    return dnsmasq_main(argc, argv);    
  else if (!strcmp (base, "arp"))
    return arp_main(argc,argv);
	const char *s;

	bb_applet_name=argv[0];
	if (*bb_applet_name == '-') bb_applet_name++;
	for (s = bb_applet_name; *s ;)
		if (*(s++) == '/') bb_applet_name = s;

	/* Set locale for everybody except `init' */
	if(ENABLE_LOCALE_SUPPORT && getpid() != 1)
		setlocale(LC_ALL, "");

	run_applet_by_name(bb_applet_name, argc, argv);
	bb_error_msg_and_die("applet not found");
}

int busybox_main(int argc, char **argv)
{
	/*
	 * This style of argument parsing doesn't scale well
	 * in the event that busybox starts wanting more --options.
	 * If someone has a cleaner approach, by all means implement it.
	 */
	if (ENABLE_FEATURE_INSTALLER && argc > 1 && !strcmp(argv[1], "--install")) {
		int use_symbolic_links = 0;
		int rc = 0;
		char *busybox;

		/* to use symlinks, or not to use symlinks... */
		if (argc > 2) {
			if ((strcmp(argv[2], "-s") == 0)) {
				use_symbolic_links = 1;
			}
		}

		/* link */
		busybox = xreadlink("/proc/self/exe");
		if (busybox) {
			install_links(busybox, use_symbolic_links);
			free(busybox);
		} else {
			rc = 1;
		}
		return rc;
	}

	/* Deal with --help.  (Also print help when called with no arguments) */

	if (argc==1 || !strcmp(argv[1],"--help") ) {
		if (argc>2) {
			run_applet_by_name(bb_applet_name=argv[2], 2, argv);
		} else {
			const struct BB_applet *a;
			int col, output_width;

			if (ENABLE_FEATURE_AUTOWIDTH) {
				/* Obtain the terminal width.  */
				get_terminal_width_height(0, &output_width, NULL);
				/* leading tab and room to wrap */
				output_width -= 20;
			} else output_width = 60;

			printf("%s\n\n"
			       "Usage: busybox [function] [arguments]...\n"
			       "   or: [function] [arguments]...\n\n"
			       "\tBusyBox is a multi-call binary that combines many common Unix\n"
			       "\tutilities into a single executable.  Most people will create a\n"
			       "\tlink to busybox for each function they wish to use and BusyBox\n"
			       "\twill act like whatever it was invoked as!\n"
			       "\nCurrently defined functions:\n", bb_msg_full_version);

			col=0;
			for(a = applets; a->name;) {
				col += printf("%s%s", (col ? ", " : "\t"), (a++)->name);
				if (col > output_width && a->name) {
					printf(",\n");
					col = 0;
				}
			}
			printf("\n\n");
			exit(0);
		}
	} else run_applet_by_name(bb_applet_name=argv[1], argc-1, argv+1);

	bb_error_msg_and_die("applet not found");
}

/*
Local Variables:
c-file-style: "linux"
c-basic-offset: 4
tab-width: 4
End:
*/
